pkg/networkservice/common/connect/server.go

```diff
@@ -16,211 +14,51 @@
-	ctx               context.Context
-	clientFactory     ClientFactory
-	clientDialTimeout time.Duration
-	clientDialOptions []grpc.DialOption
-
-	connInfos connectionInfoMap
-	clients   clientInfoMap
-	executor  multiexecutor.MultiExecutor
+	client      networkservice.NetworkServiceClient
+	callOptions []grpc.CallOption
 }
 
-type clientInfo struct {
-	client  *connectClient
-	count   int
-	onClose context.CancelFunc
-}
-
-type connectionInfo struct {
-	clientURL *url.URL
-	client    *clientInfo
-}
-
-// NewServer - server chain element that creates client subchains and requests them selecting by
-//             clienturlctx.ClientURL(ctx)
-func NewServer(
-	ctx context.Context,
-	clientFactory ClientFactory,
-	options ...Option,
-) networkservice.NetworkServiceServer {
-	s := &connectServer{
-		ctx:               ctx,
-		clientFactory:     clientFactory,
-		clientDialTimeout: time.Second,
-	}
-
-	for _, opt := range options {
-		opt(s)
+// NewServer - returns a connect chain element
+func NewServer(client networkservice.NetworkServiceClient, callOptions ...grpc.CallOption) networkservice.NetworkServiceServer {
+	return &connectServer{
+		client:      client,
+		callOptions: callOptions,
 	}
-
-	return s
 }
 
-func (s *connectServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
-	clientURL := clienturlctx.ClientURL(ctx)
-	if clientURL == nil {
-		return nil, errors.Errorf("clientURL not found for incoming connection: %+v", request.GetConnection())
-	}
-
-	c, err := s.client(ctx, request.GetConnection())
-	if err != nil {
-		return nil, err
-	}
-	if err = c.client.ctx.Err(); err != nil {
-		s.deleteClient(c, clientURL.String())
-		s.connInfos.Delete(request.GetConnection().GetId())
-		return nil, err
-	}
-
-	postponeCtxFunc := postpone.ContextWithValues(ctx)
-
-	conn, err := c.client.Request(ctx, request.Clone())
+func (c *connectServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
+	conn, err := c.client.Request(ctx, request, c.callOptions...)
 	if err != nil {
-		if _, ok := s.connInfos.Load(request.GetConnection().GetId()); !ok {
-			s.closeClient(c, clientURL.String())
-		}
 		return nil, err
 	}
-
-	// Update request.Connection
 	request.Connection = conn
-
-	_, refreshRequest := s.connInfos.Load(conn.GetId())
-	s.connInfos.Store(conn.GetId(), connectionInfo{
-		clientURL: clientURL,
-		client:    c,
-	})
-
-	// Carry on with next.Server
-	conn, err = next.Server(ctx).Request(ctx, request)
-	// Close connection if next.Server Request finished with error
-	if err != nil && !refreshRequest {
-		closeCtx, cancelClose := postponeCtxFunc()
-		defer cancelClose()
-
-		_, closeErr := s.Close(closeCtx, request.Connection.Clone())
-		if closeErr != nil {
-			err = errors.Wrapf(err, "connection closed with error: %s", closeErr.Error())
-		}
-	}
-	return conn, err
+	return next.Server(ctx).Request(ctx, request)
```

Comment:
I suppose that we most probably want to Close client chain if server Request fails, because we will return an error in such case.

Reply:
Good catch

Reply:
Fixed